test/unit/CoverProducts/migrateCoverData.js

```diff
@@ -0,0 +1,106 @@
+const { loadFixture } = require('@nomicfoundation/hardhat-network-helpers');
+const setup = require('./setup');
+
+async function migrateCoverSetup() {
+  const fixture = await loadFixture(setup);
+  const { cover, stakingPoolFactory, master, products, productTypes } = fixture;
+  await stakingPoolFactory.setStakingPoolCount(3);
+
+  const { productsList, productNames, allowedPools } = products.reduce(
+    (acc, productData) => {
+      acc.productsList.push(productData.product);
+      acc.productNames.push(productData.productName);
+      acc.allowedPools.push(productData.allowedPools);
+      return acc;
+    },
+    {
+      productsList: [],
+      productNames: [],
+      allowedPools: [],
+    },
+  );
+
+  const { productTypesList, productTypeNames } = productTypes.reduce(
+    (acc, productData) => {
+      acc.productTypesList.push(productData.productType);
+      acc.productTypeNames.push(productData.productTypeName);
+      return acc;
+    },
+    {
+      productTypesList: [],
+      productTypeNames: [],
+      allowedPools: [],
+    },
+  );
+
+  await cover.setProductsAndProductTypes(productsList, productTypesList, productNames, productTypeNames, allowedPools);
+
+  // setup new CoverProducts contract
+  const coverProducts = await ethers.deployContract('CoverProducts');
+  await coverProducts.changeMasterAddress(master.address);
+  await coverProducts.changeDependentContractAddress();
+  await master.enrollInternal(coverProducts.address);
+
+  return { ...fixture, coverProducts, productTypesList, productTypeNames, productsList, productNames, allowedPools };
+}
+
+describe('migrateCoverData', function () {
+  it('should migrate all the product, productTypes and allowedPools', async function () {
+    const fixture = await loadFixture(migrateCoverSetup);
+
+    const { coverProducts, productTypesList, productsList, allowedPools, productNames, productTypeNames } = fixture;
+
+    // migrate data
+    await coverProducts.migrateCoverData();
+
+    const productTypes = await coverProducts.getProductTypes();
+    const products = await coverProducts.getProducts();
+
+    for (const i in products) {
+      expect(products[i].productType).to.be.equal(productsList[i].productType);
+      expect(products[i].yieldTokenAddress).to.be.equal(productsList[i].yieldTokenAddress);
+      expect(products[i].coverAssets).to.be.equal(productsList[i].coverAssets);
+      expect(products[i].initialPriceRatio).to.be.equal(productsList[i].initialPriceRatio);
+      expect(products[i].capacityReductionRatio).to.be.equal(productsList[i].capacityReductionRatio);
+      expect(products[i].isDeprecated).to.be.equal(productsList[i].isDeprecated);
+      expect(products[i].useFixedPrice).to.be.equal(productsList[i].useFixedPrice);
+
+      const allowedPoolsAfter = await coverProducts.getAllowedPools(i);
+      const productName = await coverProducts.getProductName(i);
+
+      expect(allowedPoolsAfter).to.be.deep.equal(allowedPools[i]);
```

Comment:
Find a way to assert that we have:
- at least one product with 1 allowed pool
- at least one product with 2 or more allowed pools

Currently this entirely depends on the data in setup.js so it's prone to human errors

Reply:
Fork tests pass, where we have all these cases. Again I can modify the setup but like you said it's prone to human errors